src/margo-instance.h

```diff
@@ -196,6 +205,50 @@ typedef struct {
     char      is_asleep;
 } margo_thread_sleep_cb_dat;
 
+#define PROGRESS_NEEDED_INCR(__mid__)                                \
+    do {                                                             \
+        bool notify = false;                                         \
+        if ((__mid__)->progress_when_needed.flag) {                  \
+            ABT_mutex_lock(ABT_MUTEX_MEMORY_GET_HANDLE(              \
+                &(__mid__)->progress_when_needed.mutex));            \
+            notify = ++(__mid__)->progress_when_needed.pending == 1; \
+            ABT_mutex_unlock(ABT_MUTEX_MEMORY_GET_HANDLE(            \
+                &(__mid__)->progress_when_needed.mutex));            \
+        }                                                            \
+        if (notify) {                                                \
+            ABT_cond_signal(ABT_COND_MEMORY_GET_HANDLE(              \
+                &(__mid__)->progress_when_needed.cond));             \
+        }                                                            \
+    } while (0)
+
+#define PROGRESS_NEEDED_DECR(__mid__)                     \
+    do {                                                  \
+        if ((__mid__)->progress_when_needed.flag) {       \
+            ABT_mutex_lock(ABT_MUTEX_MEMORY_GET_HANDLE(   \
+                &(__mid__)->progress_when_needed.mutex)); \
+            --(__mid__)->progress_when_needed.pending;    \
+            ABT_mutex_unlock(ABT_MUTEX_MEMORY_GET_HANDLE( \
+                &(__mid__)->progress_when_needed.mutex)); \
+        }                                                 \
+    } while (0)
+
+#define WAIT_FOR_PROGRESS_TO_BE_NEEDED(__mid__)                             \
```

Comment:
Maybe these macros can be disabled (i.e. no locking etc.) if margo is initialized in listening mode?  In that case (on the server side, when unexpected requests could arrive at any time) we always have to call progress, but on clients you only need to call progress when there is margo activity.